### ISSUE TITLE
[Question] Using with agnostic service/data layer: introspection not working

### ISSUE THREAD

**Issue by webwisejordan** [nuwave/lighthouse#183]
I am using Lighthouse 2.1 in a Lumen project which has an agnostic service/data model layer (needs to support REST and GraphQL) so there is no Eloquent models to directly tie into.

In order to make this work, I created a controller for each transmission type which has generic methods for each operation (retrieve(), create(), etc.).  These all essentially work by calling a provided resolver to the service method, the main difference being that each one will catch exceptions (the service throws exceptions like ServiceCreateFailureException or ServiceUpdateFailureException among others) for the given operation and return an expected response.

I then created a custom directive, @route, which allows the schema to provide a resolver (to the service method) and operation (to point to the controller).  It works basically like @field, but defining a controller and service essentially.  This is used on every query and mutation.

So basically, the workflow is like:

incoming request --> (nuwave graphql controller) --> parses schema, @route routes request --> GraphQLController@operation --> Service@method(parameters?) --> data layer 

response --> service --> graphqlcontroller@operation --> output relevant json

This all works to run queries, mutations, etc. and everything is expected.  However, when I try to run introspection, it fails "Failed to fetch schema: Cannot read property 'types' of undefined".  When I try to add a schema node to schema.graphql, it throws an exception since the schema node is unrecognized as a valid node type in NodeFactory.

So, i'm looking for a little insight on how the introspection works within lighthouse so I can figure out how I could route introspection requests effectively, or what I might be doing that's breaking it.

Please find below, relevant code examples for further context.  Thank you for all your great work and for your help on this issue!

Schemas
```
# schema.graphql

#import user.graphql
type Query { }
type Mutation { }
type ResponsePayload {
  status: String!
  data: Object
}

# user.graphql

extend type Query {
    users: [User!]! @route(resolver: "User@getAll" operation: "retrieve")
    user(id: Int!): User! @route(resolver: "User@get" operation: "retrieve")
}

extend type Mutation {
    createUser(
        email: String!
        prefix: String
        firstName: String!
        middleName: String
        lastName: String!
        suffix: String
    ): ResponsePayload
    @route(resolver: "User@create" operation: "create")
}
```
Route directive
```
class RouteDirective implements FieldResolver
{
    use HandlesDirectives;

    /**
     * @var string The controller used to pass through service requests
     */
    private const CONTROLLER = '\\App\\Http\\Controllers\\GraphQLController';
    /**
     * @var string The field resolver
     */
    protected $resolver;
    /**
     * @var string The field resolver method
     */
    protected $method;

    /**
     * Name of the directive.
     *
     * @return string
     */
    public function name()
    {
        return 'route';
    }

    /**
     * Resolve the field directive.
     *
     * @param \Nuwave\Lighthouse\Schema\Values\FieldValue The field to which this route directive refers
     * @return \Nuwave\Lighthouse\Schema\Values\FieldValue Updated field value with resolver set by given parameters
     */
    public function resolveField(FieldValue $value)
    {
        // get the directive node information
        $directive = $this->fieldDirective($value->getField(), $this->name());

        // get the resolver and operation
        $resolver = $this->getResolver($directive);
        $operation = $this->getOperation($directive);

        return $value->setResolver(
            // note: although root and context are not referenced here, they are NECESSARY for lighthouses internal routing
            function ($root, array $args, $context = null, $info = null) use ($directive, $resolver, $operation) {
                // set the resolver to context information and verify the controller route is valid
                $info->serviceResolver = $resolver;
                $reflect = new \ReflectionClass($this::CONTROLLER);
                if (!$reflect->isInstantiable()) {
                    // this is a config error in the code rather than a schema issue so throw a http internal 500 error
                    throw new HttpException('GraphQL Controller does not exist.');
                }
                if (!$reflect->hasMethod($operation)) {
                    throw new DirectiveException(sprintf(
                        'Directive [%s] `operation` argument is invalid. Method does not exist in controller.',
                        $directive->name->value
                    ));
                }

                // instantiate the controller and call the routed operation
                try {
                    $controller = $reflect->newInstance($resolver, $args, $info);
                    return $controller->$operation();
                } catch (BadRouteException $e) {
                    throw new DirectiveException(sprintf(
                        'Directive [%s] `resolver` argument is invalid. ' . $e->getMessage(),
                        $directive->name->value
                    ));
                }
            }
        );
    }

    /**
     * Get resolver for route.
     *
     * @param \GraphQL\Language\AST\DirectiveNode The node from the route directive
     * @throws \Nuwave\Lighthouse\Support\Exceptions\DirectiveException If the resolver doesn't exist in the directive
     * @return string
     */
    protected function getResolver(DirectiveNode $directive)
    {
        $resolver = $this->directiveArgValue($directive, 'resolver');

        if (!$resolver) {
            throw new DirectiveException(sprintf(
                'Directive [%s] must have an `resolver` argument.',
                $directive->name->value
            ));
        }

        return $resolver;
    }

    /**
     * Get operation for route.
     *
     * @param \GraphQL\Language\AST\DirectiveNode The node from the route directive
     * @throws \Nuwave\Lighthouse\Support\Exceptions\DirectiveException If the operation doesn't exist in the directive
     * @return string
     */
    protected function getOperation(DirectiveNode $directive)
    {
        $operation = $this->directiveArgValue($directive, 'operation');

        if (!$operation) {
            throw new DirectiveException(sprintf(
                'Directive [%s] must have an `operation` argument.',
                $directive->name->value
            ));
        }

        return $operation;
    }
}
```

constructor and representative methods of graphql controller
```
    /**
     * Initialize the state of the request
     *
     * @param \Dingo\Api\Http\Request The request to process
     * @param array The arguments which are passed to the operation
     * @param \GraphQL\Type\Definition\ResolveInfo Information about the resolver
     */
    public function __construct($resolver, array $args, $info)
    {
        // ensure that there is a resolver set in the route and that it is valid
        if (!$resolver) {
            throw new BadRouteException('Missing resolver in route.');
        }
        if (strpos($resolver, '@') === false) {
            throw new BadRouteException('Malformed resolver in route.');
        }

        // set model and method from resolver and verify that they are valid
        list($this->serviceName, $this->methodName) = explode('@', $resolver);
        $this->serviceName .= 'Service';

        // valid resolvers point to services which are instantiable subclasses of BaseService in the Services namespace
        $reflect = new \ReflectionClass($this::SERVICE_NAMESPACE . $this->serviceName);
        if (!$reflect->isInstantiable()
            || !$reflect->isSubclassOf($this::SERVICE_NAMESPACE . 'BaseService')
            || !$reflect->inNamespace()
            || '\\' . $reflect->getNamespaceName() . '\\' != $this::SERVICE_NAMESPACE
        ) {
            throw new BadRouteException('Resolver points to invalid service.');
        }
        if (!$reflect->hasMethod($this->methodName)) {
            throw new BadRouteException('Resolver points to method which does not exist.');
        }
        $this->service = $reflect->newInstance();

        // define parameters and the fields which should be returned
        $this->parameters = $args;
        $this->returnFields = $info->getFieldSelection($this::RETURN_FIELD_DEPTH);
        $this->operation = $info->operation->operation;
        $this->returnType = $info->returnType;
    }

    /**
     * Process a retrieval query
     *
     * @return void
     */
    public function retrieve()
    {
        try {
            $this->issueResponse(
                $this->service->{$this->methodName}($this->parameters)
            );
        } catch (ServiceEntityNotFoundException $e) {
            $this->handleException($e->getMessage());
        }
    }

    /**
     * Process a create mutation
     *
     * @return void
     */
    public function create()
    {
        try {
            $entity = $this->service->{$this->methodName}($this->parameters);
            $this->issueResponse(
                $this->convertEntityToArray($entity)
            );
        } catch (ServiceValidationFailureException $e) {
            $this->handleException($e->getMessage(), $e->getErrors());
        } catch (ServiceCreateFailureException $e) {
            $this->handleException($e->getMessage());
        } catch (\Exception $e) {
            $this->handleException($e->getMessage());
        }
    }

    /**
     * Handles exception states and issues a properly formatted GraphQL error response
     *
     * @param String The overall error message
     * @param \Illuminate\Support\MessageBag The list of validation errors returned from the action, if any
     * @return void
     */
    private function handleException($message, MessageBag $errors = null)
    {
        $errorList = [
            (Object)[
                "message" => $message
            ]
        ];
        if (!is_null($errors) && $errors instanceof MessageBag) {
            foreach ($errors->getMessages() as $field => $messages) {
                foreach ($messages as $msg) {
                    array_push($errorList, (Object)[
                        "message" => $msg,
                        "path" => [$field] // TODO: Update this to support full path to schema field
                    ]);
                }
            }
        }

        $this->issueResponse(null, $errorList);
    }

    /**
     * Issues a properly formatted GraphQL response
     *
     * @param array|null The data to return in the response
     * @param Object[] A list of errors which exist for this operation
     * @return void
     */
    private function issueResponse($data = null, array $errorList = [])
    {
        // if there are errors, the query data is ignored and we return an error response
        if (is_array($errorList) && sizeof($errorList) > 0) {
            echo json_encode([
                'success' => false,
                'data' => null,
                'errors' => $errorList
            ]);
            die();
        }

        // build full response object, then trim to match return fields structure
        $response = [];
        if ($this->operation == 'mutation') {
            $response = $this->mapDataToFieldSet([
                "success" => true,
                "data" => $data
            ], $this->returnFields);
        } elseif ($this->operation == 'query') {
            $response = $this->mapDataToFieldSet($data, $this->returnFields);
        }
        header('Content-Type: application/json');
        echo json_encode($response);
        die();
    }
```

**Comment by chrissm79** [comment]
@webwisejordan Pretty interesting use-case!! I might not be able to dig into this until this afternoon but just taking a quick look, did you define an `Object` scalar type (used in the `ResponsePayload` type)?

**Comment by webwisejordan** [comment]
Actually, I didn't define a scalar type for Object.  I was thinking it was one of the native types for some reason, but a quick check against the spec confirmed that it was not.  I will make sure to update it when I get home.

Strangely though it doesn't give any kind of schema validation error, and the queries and mutations still work as expected.  I did try some other tests, including just having one query and deleting everything else (including the ResponsePayload type) and still encountered the same result on introspection.

**Comment by chrissm79** [comment]
@webwisejordan Yeah, I'm not sure why your queries/mutations would still work but I know I personally ran into that error and I *believe* it was because I was extending a `interface` that I didn't actually define. I might need to dig into that to see if that's a Lighthouse or `graphql-php` issue (I would think you should get a schema error as well).

Let me know if defining the Object type (or using some sort of interface/union) fixes your issue or if you're still running into the problem!

**Comment by spawnia** [comment]
@webwisejordan Well actually, the schema is not fully validated during normal requests. For performance reasons, the types are only loaded lazily. If you want to debug your schema and ensure it is valid, you can use `php artisan lighthouse:validate-schema`

**Comment by webwisejordan** [comment]
@chrissm79 I just was able to try and neither defining a scalar for Object nor just replacing it with a String worked - same result.

@spawnia Interesting, and definitely good to know.  I tried to `php artisan lighthouse:validate-schema` and it said it was undefined.  I am using Lumen rather than Laravel, so perhaps that has something to do with it.

**Comment by chrissm79** [comment]
@webwisejordan If the schema is currently being cached you probably want to set the option to false for the time being as changes while you debug won't take effect. Other than that it shouldn't haven't any effect on this issue. 

As I recall, I had a misspelled field type and it caused the same error. Can you remove all your imports and just import your user file and comment out all fields from the user type except the email. You can also keep the user and users fields on the query your to test with and let me know if the introspection works with that. That should tell us if it's something with your schema or with the code.

**Comment by spawnia** [comment]
@webwisejordan Right, sorry for that. The commands changed in v2.2, they are not available in v2.1

There are some major performance improvements in v2.2, so if you are feeling lucky, maybe try `dev-master`

**Comment by webwisejordan** [comment]
@spawnia  Good to know.  I will wait until I have the other issues fixed and possibly until it is pushed to stable - but I'll definitely give it a go.

@chrissm79 So, I played around a lot more with schema manipulations, starting with just a bare Query and Mutation, I could see things showing up! :D However, it seems that the issue was with my scalar definition for DateTime - as soon as I uncomment that, it breaks again (even if nothing references it).

I've looked several times and can't see anything wrong, but maybe I'm missing something simple.  Here is the relevant code:

schema.graphql
```scalar DateTime @scalar(class: "DateTime")```

App\Http\GraphQL\Scalars\DateTime.php
```
<?php namespace App\Http\GraphQL\Scalars;

use Nuwave\Lighthouse\Schema\Types\Scalars\DateTime as LightDateTime;

class DateTime extends LightDateTime
{
}
```
namespace registry in config
```
    'namespaces' => [
        'models' => 'App\\Models',
        'mutations' => 'App\\Http\\GraphQL\\Mutations',
        'queries' => 'App\\Http\\GraphQL\\Queries',
        'scalars' => 'App\\Http\\GraphQL\\Scalars',
    ],
```

**Comment by chrissm79** [comment]
@webwisejordan Do me a favor and use the full namespace in your scalar definition like so:

```graphql
scalar DateTime @scalar(class: "App\\Http\\GraphQL\\Scalars\\DateTime")
```

Let me know if that works, it might be that we're not resolving the namespace in the scalar directive and it's using `\DateTime` instead of your scalar.

**Comment by spawnia** [comment]
@chrissm79 Could it be a conflict between the DateTime scalar we already have in Lighthouse?

**Comment by webwisejordan** [comment]
@chrissm79 I did try the full namespace as well with the same result.  Could be a conflict as @spawnia suggested I suppose.  On a sidenote, is there anywhere that has a list of the lighthouse-provided Scalars available by default in documentation?

**Comment by chrissm79** [comment]
@webwisejordan I'm still not sure why you wouldn't be able to extend the lighthouse scalar, I'm able to do so on a fresh install, but it's a Laravel build not Lumen. 

However, I did notice that if you defined the scalar w/ the entire namespace it still prepended the namespace defined in the lighthouse config. I pushed a small patch that will fallback to the `class` defined in the schema and if it can't find the class at all it will throw an error with a more helpful message.
